### DISCUSSION TITLE
Immersive Menus, Interactions, Items, ETC

### DISCUSSION THREAD

**Discussion by LPlusBonk** [QuestCraftPlusPlus/QuestCraft#273] [Ideas]
Hello! First off, love how it's going so far, yall are doing great work. I understand how early in development Questcraft is currently, but I couldn't help sharing a few ideas for future feature improvements.

**By the way, most of these are entirely unrealistic performance-wise and also because you are making a mod, not an entirely new game, but I hope it can give you some inspiration!**

Also, some of these may be conflicting, its just brainstorming

### Basic
Here are some really simple ideas with a very small impact on the current game, and are entirely unrelated to the other ideas:
- Show crosshair on menus or a ray from the dominant controller
- Switch between Hotbar slots by pointing with the dominant hand at chosen slot
- 1.18 caves + cliffs
- Make block damage stay around for a while longer, so that manual interactions are more possible later on
- Build in the performance-enhancing modifications and an LOD mod
- Patch clipping camera through walls
- Performance
- Movement should be around hand, not head

### Player
Remember Minecraft blender animations and the trailers?
- Legs could have an IK if performance, unlikely though
- Arms could have the forearms attached to the controller loosely, with a maximum extended length
- Arms could have the upper arm attached to the shoulder on one end and smoothly deform into the forearm on the other end to make a single continuous arm with an elbow
- Held items' rotation could be strongly attached to the rotation of the controller for greater control, like a wrist that doesn't need to deform the player model anymore
- Held items' position could be strongly attached to the end of the forearm to prevent reach exploitation
- Held items could be placed more accurately to match the user's hands
- Sleeping in a bed places your camera above the bed as if you are standing on it, but puts your player model on it correctly, to prevent motion sickness and be cool
- 3D proximity voice chat

### 2-Handed
Cmon, in real life you can hold two tools, they just both work less!
- No dominant hand. Both hands are equally useful, they both have a trigger and a grip, that's all they need
- They each have a crosshair
- Grip the same tool or weapon with two hands to make it more effective
- Using two pickaxes of the different levels doesn't round to the dominant hand, each stroke just does the damage of the pickaxe used
- Items like bows and crossbows need both hands

### Hotbar
This isn't a flatscreen game anymore, so it shouldn't have a UI like one!
- Display the Hotbar on both arms, so you can pull items out with either hand
- Don't show the Hotbars all the time, only when pointing or looking at one or the other
- Show the health and chat above one, and the hunger and XP levels above the other These should auto-hide too, except for when one changes, when the hunger or health gets low, or when something is sent in chat. These events should only show the relevant one
- Press and hold the grip button while pointing at or touching the targeted Hotbar slot to pull out an item into that hand
- Pulling an item out makes that slot now empty
- Pulling parts of a stack of items out should be possible, so you can hand a few to someone else

### Items
Everyone's favorite thing to do in VR is throwing things around
- Make items become similar to the item physic mod, except sharp things which can stick to surfaces
- Pick items up by pointing at them and holding the grip button to bring them into that hand
- Releasing the grip button drops the item
- Dropped items inherit the inertia of your controller when released, so you can throw them
- Release an item near your hip to send it to your inventory
- Release an item while touching or pointing at an inventory slot to place it back in
- Trading items, grabbing an item someone else is holding, followed by them releasing their grip on it without pointing or touching their Hotbar or hip gives you the item

### Tools and weapons
I wish I could just close my index finger around a vacuum and it would clean for me, but I guess that's not how it works
- Taking out a 2 handed weapon or tool with one hand should show a new slot only on the Hotbar attached to the arm holding the tool that allows your other hand to grab the other part. For instance:
  - A bow should show slots for all of the types of arrows you have, allowing you to pull out the kind you want
  - A flint and steel should show the steel
  - A crossbow should show arrows and rockets
  - Idk what else, but it should be fairly obvious

#### Tools
- A hoe could automatically re-plant the seeds when swung horizontally, but not when swinging vertically
- When the hoe is swung with two hands, it could affect all hit crops further away, not just the ones the model passes through
- When the head of an axe or pickaxe makes contact with something, it should not just travel through freely, but stop at a certain depth, and need to be pulled back out to do more damage
- Using two hands with an axe or a pickaxe does more damage per swing
- If you release an ax or pickaxe while it is in a block, it remains there
- Shovels should be able to slide through blocks that can be mined with shovels, but with resistance. On non-shovel blocks, the shovel should behave like the other tools
- Swinging tools underwater has more resistance
- Flint and steel should be clinked together
- Sheers should operate on trigger click
- Fishing rod uses one hand to toss and hold, but needs another hand to reel in

#### Weapons
- For swords and axes, swinging down vertically should do crits
- Using two hands with swords and axes does more damage, but you can't use a shield
- Tridents could be thrown like an item, see thrown subsection
- Bows and crossbows need to be drawn, crossbows can be stored drawn. Draw with the trigger on hand holding an arrow, and shoot the crossbow by clicking the trigger
- Shields shouldn't have a block action, but should always be in effect, only affecting attacks coming from perpendicular to its face

#### Thrown
- Swords and axes can be thrown. Doing so damages mobs they pass through and sticks into the block they hit
- Axes, pickaxes, and shovels can damage blocks they hit when thrown. If they hit the underside of a block, they fall back down
- Blocks damaged by throwing remain damaged for longer than usual, allowing another throw to take place
- All tools can be impaled in the ground and stay there if contacted by the head
- Tridents don't get slowed in water, others do

### Armor
- Place armor on yourself

### Inventory
- Releasing an item near your hip sends it to your inventory
- Make the inventory a physical page that can be freely moved around, and not seen by other players, toss away to close
- Remove the crafting grid, the Hotbar section, and the player view, and make the armor slots not changeable, so that the only purpose they serve is showing the enchantments and durability of the equipped armor
- Opening a chest by lifting its lid shows its inventory. Alternatively, if the chest is far away from the player, a physical inventory menu pops out with a line connecting it to its chest. You can open up to 5 chests at once, but they will close if you move out of range or open another (only one at a time, in this case, keeping only 5 open max)
- Moving items between slots is more like taking the item out of one inventory and putting it in another
- Furnaces show one slot on the bottom slot for fuel, and two slots on the top slot for the item and product
- To pick up an item, use the grip button
- To pick up half, hold the trigger, then the grip button, then release the trigger
- To distribute items evenly from a slot, hold the trigger and drag
- To release one item in your hand, click the trigger on an empty slot
- To quick transfer an item, grip and toss to a different inventory

### Crafting
- Uses the texture on the top of the block
- Recipe book button and menu on the left side of the grid
- Finished item on the right side of the grid
- Grabbing a recipe pulls the required items into your hand so you can tell what it will be taking
- Placing a recipe onto the grid places down the required items, and shows the recipe above the grid
- When a recipe is active, the grid is outlined in green
- Picking up a product moves new required items from your inventory onto the grid to craft again. Removing the recipe above the grid and tossing it away sends the items back into your inventory
- pressing the trigger on the product crafts more, if its not stackable or you reached the maximum stack limit, it sends the product to your inventory

### Riding
- Reins for horses and donkeys
- Rowing for boats
- Swimming for water
- push the ground for minecarts

### Chat
- Physical menu with a physical keyboard that you can press with your hands, toss away to close

### Menus
- Physical, toss away to close

### Launcher
- Built-in launcher with all performance mods set up automatically

### Experimental ideas
Remember the 2022 April Fools Minecraft update?
- One block at a time plus
  - No inventory or Hotbar
  - Tools and items can be hung on the back, hips, and belt
  - You can pick up and carry a chest with 2 hands. If you have the strength effect, you can carry a double chest
  - 2x2 crafting in the palm of your hand
- PVP could have persistent weapon interaction, IE they collide with each other
- Attacks could reveal a dutch angle red on black grid like in SAO in place of the normal texture in place of contact

**Thanks for reading!** Hope you didn't have anything you planned on doing though, sorry about that :)
***\\/ Like if you read all the way here!***

**Comment by ghost** [comment]
Y e s  This is kinda my what I would love to be in Minecraft vr :D :D

**Comment by wellsilver** [comment]
Bro!! Copy and paste to MCXR the changes here are only really to make it compatible for quest and do better on it

**Comment by Cartrigger** [comment]
Really good ideas, but don't forget to stay to Vanilla.
